File: 101-250/234-palindrome-linked-list.cpp
bool isPalindrome(ListNode* head) {
    
    stack<int> s;
    ListNode *curr = head;
    while (curr) {
        s.push(curr->val);
        curr = curr->next;
    }

    curr = head;
    while (curr){
        if (curr->val == s.top()){
            s.pop();
            curr = curr->next;
        }

        else break;
    }

    return s.empty();
}

//divide in half and reverse

bool isPalindrome(ListNode* head) {
    if (!head->next) return true;
    ListNode *slow = head, *fast = head;
    while (fast->next && fast->next->next) {
        slow = slow->next;
        fast = fast->next->next;
    }
    slow->next = reverseList(slow->next);
    slow = slow->next;
    while (slow) {
        if (head->val != slow->val) {
            return false;
        }
        head = head->next;
        slow = slow->next;
    }
    return true;

}

ListNode* reverseList(ListNode* head) {
    ListNode *prev = nullptr, *next;
    while (head) {
        next = head->next;
        head->next = prev;
        prev = head;
        head = next;
    }
    return prev;
}

File: python version/234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head.next: return True

        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        halfway = slow
        
        halfway.next = self.reverseList(halfway.next)
        first_start, second_start = head, halfway.next

        while second_start:
            if first_start.val != second_start.val:
                return False
            else:
                first_start = first_start.next
                second_start = second_start.next

        return True


    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        while curr:
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp
        return prev